cmd/up/space/defaults/defaults.go

```
@@ -15,74 +15,104 @@
-	d := map[string]string{
-		"clusterType": string(cloud),
+func (ct *CloudType) getSpaceValues() map[string]string {
+	if v, ok := vendorDefaults[*ct]; ok {
+		return v
 	}
+	return nil
+}
 
-	for k, v := range d {
-		add := true
-		for cs := range s {
-			if cs == k {
-				add = false
-				break
-			}
-		}
-		if add {
-			s[k] = v
-		}
+func (ct *CloudType) Defaults() CloudConfig {
+	publicIngress := true
+	if *ct == Generic || *ct == Kind {
+		publicIngress = false
+	}
+	return CloudConfig{
+		SpacesValues:  ct.getSpaceValues(),
+		PublicIngress: publicIngress,
 	}
-	return s, nil
 }
 
-func detectKubernetes(kClient kubernetes.Interface) (CloudType, error) {
-	ver, err := kClient.Discovery().ServerVersion()
+func GetConfig(kClient kubernetes.Interface) (*CloudConfig, error) {
+	if kClient == nil {
+		return nil, errors.New("no kubernetes client")
+	}
+	cloud, err := detectKubernetes(kClient)
 	if err != nil {
-		return Generic, err
+		return nil, err
 	}
 
-	// GKE has -gke in the git commit
-	// Example:
-	if strings.Contains(ver.GitVersion, "-gke.") {
-		return GoogleGKE, nil
+	if cloud == Generic || cloud == Kind {
+		pterm.Info.Printfln("Setting defaults for vanilla Kubernetes (type %s)", string(cloud))
+		return nil, nil
 	}
-	// EKS has -eks in the git commit
-	if strings.Contains(ver.GitVersion, "-eks-") {
-		return AmazonEKS, nil
-	}
-	// AKS has -aks in the git commit
-	if strings.Contains(ver.GitVersion, "-aks") {
-		return AzureAKS, nil
+
+	pterm.Info.Printfln("Applying settings for Managed Kubernetes on %s", strings.ToUpper(string(cloud)))
+	return &CloudConfig{
+		SpacesValues:  cloud.Defaults().SpacesValues,
+		PublicIngress: cloud.Defaults().PublicIngress,
+	}, nil
+}
+
+func detectKubernetes(kClient kubernetes.Interface) (CloudType, error) {
+	// EKS and Kind are _harder_ to detect based on version, so look at node labels.
+	ctx := context.Background()
+	if nodes, err := kClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{}); err == nil {
```

Do we want to return the error if it's non-nil?

---

Maybe? I didn't error since nodes is not a required aspect of Spaces. If the detection fails, I didn't think that this worthy to abort the installation since the cloud type could be set via `--set clusterType=`

---

OK. If we don't think it's worthy, maybe we could add a comment for why we don't care and then handle the lint err 👍